modules/libcom/src/osi/os/posix/osdMemFile.c

```diff
@@ -0,0 +1,125 @@
+/*************************************************************************\
+* Copyright (c) 2022 Michael Davidsaver
+* SPDX-License-Identifier: EPICS
+* EPICS BASE is distributed subject to a Software License Agreement found
+* in file LICENSE that is included with this distribution.
+\*************************************************************************/
+
+#if defined(__linux__)
+#  ifndef _GNU_SOURCE
+#    define _GNU_SOURCE
+#  endif
+#  define USE_FOPENCOOKIE
+
+#else
+/* freebsd, OSX, RTEMS */
+#  define USE_FUNOPEN
+#endif
+
+#include <stdlib.h>
+#include <string.h>
+#include <errno.h>
+
+#include "epicsStdio.h"
+#include "epicsMemFilePvt.h"
+
+#if defined(USE_FOPENCOOKIE)
+typedef size_t loc_t;
+typedef ssize_t sloc_t;
+typedef cookie_io_functions_t io_functions_t;
+
+#define xfopen(COOKIE, FUNCS) fopencookie(COOKIE, "r", FUNCS)
+
+#elif defined(USE_FUNOPEN)
+typedef fpos_t seek_off_t;
+typedef int loc_t;
+typedef int sloc_t;
+
+typedef struct {
+    int (*close)(void *raw);
+    int (*read)(void *raw, char* buf, loc_t cnt);
+} io_functions_t;
+
+#define xfopen(COOKIE, FUNCS) funopen(COOKIE, (FUNCS).read, NULL, NULL, (FUNCS).close)
+#endif
+
+typedef struct {
+    const epicsIMF* mf;
+    size_t pos;
+    const char* actual;
+} mfCookieActual;
+
+static
+int mfCloseActual(void *raw)
```

Comment:
I don't get this naming. Why not just `mfCookie`, `mfClose` etc?

Reply:
"Actual" refers to `mfCookieActual::actual` pointing to uncompressed data.  At one point I was thinking about having a different set of callback functions to handle uncompressed vs. compressed.  With the compressed case doing incremental de-compression.  For example, to avoid needing to allocate a 512kB temporary buffer for `softIoc.dbd`.  So far, I've talked myself out of what seems like a premature optimization.

Reply:
Regarding incremental decompression, I agree.

As for the naming, there's two separate issues. One is that the *Actual* part of `mfCloseActual` and `mfCookieActual` doesn't add anything of value; a cookie is a cookie, there is nothing "actual" about it. The other issue is that "actual" in `epicsIMF::actualLen` *does* make sense in that context as it contrasts with `epicsIMF::contentLen`. Here, not so much. I think it would be better to replace "actual" with "uncompressed", both here and in `epicsIMF`. `EPICS_MEM_FILE_RAW` is just a trivial compression algo, after all.

Reply:
I've dropped "Actual" from these names.